### PR TITLE
add log context management to log.Context and log.RateLimtedLogger

### PR DESCRIPTION
**Pull request by codesmith14**

### CODE REVIEW

File: log/ratelimit.go [hunk 1]
```diff
@@ -25,15 +25,10 @@ func (r *RateLimitedLogger) now() time.Time {
...
-	now := r.now()
-	if r.EventCounter.Event(now) > r.Limit {
-		if r.LimitLogger != nil {
-			// Note: Log here messes up "caller" :/
-			r.LimitLogger.Log(kvs...)
-		}
-		return
+	// make the log statement
+	if logCtx := r.Context(); logCtx != nil && !IsDisabled(logCtx) {
+		logCtx.Logger.Log(addArrays(copyIfDynamic(logCtx.KeyVals), kvs)...)
```
Comment (mdubbyap) [inline log/ratelimit.go]: have we done any benchmarks on this?

**Comment by charliesignalfx** [comment]
@mdubbyap I added benchmarks to master.  Ran them.  Rebsed this PR.  Made a few tweaks to this pr.  And... ran the benchmarks again.

```
RateLimitedWithContext

before:
BenchmarkRateLimitedWithContext-12    	   41260	     28686 ns/op	    4320 B/op	      60 allocs/op

after:
BenchmarkRateLimitedWithContext-12    	   39378	     29975 ns/op	    4000 B/op	      50 allocs/op
RateLimitedWithoutContext

before:
BenchmarkRateLimitedWithOutContext-12    	  165026	      6429 ns/op	     160 B/op	      10 allocs/op

after:
BenchmarkRateLimitedWithOutContext-12    	  126037	      9301 ns/op	     640 B/op	      20 allocs/op
```

File: log/log.go [hunk 2]
```diff
@@ -135,34 +146,50 @@ func (f LoggerFunc) Log(keyvals ...interface{}) {
 	f(keyvals...)
 }
 
+func getContext(l Logger) *Context {
+	if logger, ok := l.(LoggerWithContext); ok {
```
Comment (mdubbyap) [inline log/log.go]: The benchmark numbers for rate limited seem high.  I think this is returning NewContext when called under those circumstances. Can we not do the nil check i call out below? I feel like this is going to save us some calls. But the benchmarks will tell us.  Can you run that test with pprof and see where the allocations are that basically doubled?

File: log/ratelimit.go [hunk 3]
```diff
@@ -44,6 +44,17 @@ func (r *RateLimitedLogger) Disabled() bool {
 	return r.EventCounter.Events(now, 0) > r.Limit || IsDisabled(r.Logger)
 }
 
+// Context returns the underlying log context based on the rate limit
+func (r *RateLimitedLogger) Context() *Context {
+	// get the appropriate logger from the ratelimited logger
+	if r.EventCounter.Event(r.now()) > r.Limit {
+		return getContext(r.LimitLogger)
```
Comment (mdubbyap) [inline log/ratelimit.go]: i think the lack of the nil check here is causing us issues.

**Pull request closed without merging**